### PR TITLE
Refactor DesignToggleRow

### PR DESCRIPTION
**Pull request by islemaster**
_Pair: @bcjordan_ 

**Note: This change is pure refactor, delta one fix, described below.**

Refactor the `DesignToggleRow` component.  It's now named `PlaySpaceHeader` and composed of several child components so that we can reuse them in upcoming work:
- PlaySpaceHeader
  - ToggleGroup
    - ToggleButton
  - ScreenSelector
  - ViewDataButton

The one fix we're doing here was a subtle "jump" behavior in the "Design" mode button when you clicked on the "Code" mode button, due to an `!important` CSS border-width rule taking effect in the `:active` state of the "Code" mode button.

Benefits to this refactor:
- The `ToggleGroup` component is now easier to reuse in other places, or to reconfigure for any number of buttons (not hard-coded to two buttons as it is now).  This is important as upcoming GameLab work will change the buttons to "Code | Animation" and upcoming MakerLab work will add a third button "Code | Design | Simulator".
- Smaller, easier-to-understand components.
- Slightly less rendering - several elements are simply not rendered instead of being needlessly rendered with 'display:none'.  We were careful to verify that no JavaScript depended on these elements being invisibly available in the DOM.
- Introduced a tiny starter test for the `ToggleButton` component.

### CODE REVIEW

File: apps/src/applab/PlaySpaceHeaderStyles.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  buttonStyle: {
+    display: 'inline-block',
+    verticalAlign: 'top',
+    border: '1px solid #949ca2',
+    margin: '0 0 8px 0',
+    padding: '2px 6px',
+    fontSize: 14
+  },
+  toggleButtonStyle: {
+    borderRightWidth: '0 !important',
```
Comment (islemaster) [inline apps/src/applab/PlaySpaceHeaderStyles.js:19]: Making this rule `!important` fixes the one-pixel jump when clicking the "Code" button, which is caused by a 1px !important top-level border style on buttons in common.css.

File: apps/src/applab/designMode.js [hunk 2]
```diff
@@ -859,7 +859,7 @@ designMode.renderToggleRow = function (screenIds) {
   var designToggleRow = document.getElementById('designToggleRow');
   if (designToggleRow) {
     React.render(
-      React.createElement(DesignToggleRow, {
+      React.createElement(PlaySpaceHeader, {
```
Comment (Bjvanminnen) [inline apps/src/applab/designMode.js:862]: This comment isn't particular to this PR, and not something you necessarily need to address. But I think with our current build system we can using JSX in .js files. This means, we could get rid of the React.createElement here and do

```
<PlaySpaceHeader
  hideToggle: {Applab.hideDesignModeToggle()}
  ...
/>
```

Comment (islemaster) [inline apps/src/applab/designMode.js:862]: What?  That's awesome.  I had no idea.

On Mon, Feb 22, 2016 at 4:44 PM, Brent Van Minnen notifications@github.com
wrote:

> In apps/src/applab/designMode.js
> https://github.com/code-dot-org/code-dot-org/pull/6916#discussion_r53720039
> :
> 
> > @@ -859,7 +859,7 @@ designMode.renderToggleRow = function (screenIds) {
> >    var designToggleRow = document.getElementById('designToggleRow');
> >    if (designToggleRow) {
> >      React.render(
> > -      React.createElement(DesignToggleRow, {
> > -      React.createElement(PlaySpaceHeader, {
> 
> This comment isn't particular to this PR, and not something you
> necessarily need to address. But I think with our current build system we
> can using JSX in .js files. This means, we could get rid of the
> React.createElement here and do
> 
> <PlaySpaceHeader
>   hideToggle: {Applab.hideDesignModeToggle()}
>   ...
> />
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/6916/files#r53720039.

## 

Brad Buchanan
Software Engineer, Code.org
brad@code.org

Comment (Bjvanminnen) [inline apps/src/applab/designMode.js:862]: I'm not 100% on that, but I am like 80% or so.

File: apps/src/applab/constants.js [hunk 3]
```diff
@@ -1,5 +1,9 @@
 module.exports = {
   FOOTER_HEIGHT: 30,
   DESIGN_ELEMENT_ID_PREFIX: 'design_',
-  NEW_SCREEN: "New screen..."
+  NEW_SCREEN: "New screen...",
+  MODE: {
+    CODE: 'CODE',
+    DESIGN: 'DESIGN'
```
Comment (Bjvanminnen) [inline apps/src/applab/constants.js:7]: Are these strings user visible? If so, I'd prefer they live in our i18n pathway. If this is just an enum, this is fine.

File: apps/src/applab/ViewDataButton.jsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ */
+var ViewDataButton = React.createClass({
+  propTypes: {
+    onClick: React.PropTypes.func.isRequired
+  },
+
+  render: function () {
+
+    var showDataButtonStyle = $.extend(
+        {
```
Comment (Bjvanminnen) [inline apps/src/applab/ViewDataButton.jsx]: should this be a two space indent instead of 4?

File: apps/src/applab/ViewDataButton.jsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * The button above the visualization that opens the data browser.
+ * @type {function}
+ */
+var ViewDataButton = React.createClass({
+  propTypes: {
+    onClick: React.PropTypes.func.isRequired
+  },
+
+  render: function () {
+
```
Comment (Bjvanminnen) [inline apps/src/applab/ViewDataButton.jsx]: I've been seeing various folks stick an empty line at the top of functions in a bunch of places recently. Any reason for this? Not a big deal really, just wondering if there's a justification. I personally don't like it, but can live with it :-P

File: apps/src/applab/ViewDataButton.jsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        {
+          float: 'right',
+          textAlign: 'left',
+          maxWidth: '100%',
+          whiteSpace: 'nowrap',
+          overflow: 'hidden',
+          textOverflow: 'ellipsis'
+        },
+        styles.buttonStyle,
+        styles.inactiveStyle
```
Comment (Bjvanminnen) [inline apps/src/applab/ViewDataButton.jsx]: Feels a little weird that we let buttonStyle/inactiveStyle potentially override us? Would this make more sense as 

```
$.extend({}, 
  styles.buttonStyle,
  styles.inactiveStyle
  {
    float: 'right'
    ...
  }
});
```

File: apps/src/applab/ToggleGroup.jsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    return <span>{this.renderChildren()}</span>;
+  },
+
+  renderChildren: function () {
+    return React.Children.map(this.props.children, function (child, index) {
+      return React.cloneElement(child, {
+        first: index === 0,
+        last: index === React.Children.count(this.props.children) - 1,
+        active: child.props.value === this.props.selected,
+        onClick: this.setSelected.bind(this, child.props.value)
```
Comment (Bjvanminnen) [inline apps/src/applab/ToggleGroup.jsx]: We could instead do:

```
this.props.onChange.bind(this, child.props.value)
```

Right? And then no longer need this.setSelected

File: apps/src/applab/ToggleButton.jsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    id: React.PropTypes.string,
+    active: React.PropTypes.bool.isRequired,
+    first: React.PropTypes.bool,
+    last: React.PropTypes.bool,
+    onClick: React.PropTypes.func.isRequired
+  },
+
+  render: function () {
+    return (
+        <button
```
Comment (Bjvanminnen) [inline apps/src/applab/ToggleButton.jsx]: extra indentation

File: apps/src/applab/ToggleButton.jsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+            className='no-outline'
+            onClick={this.props.onClick}>
+          {this.props.children}
+        </button>
+    );
+  },
+
+  getStyle: function () {
+    return $.extend({},
+        styles.buttonStyle,
```
Comment (Bjvanminnen) [inline apps/src/applab/ToggleButton.jsx]: extra indentation (i think. i know there are times where we consider 4 spaces valid, but i dont _think_ this is one?)

File: apps/src/applab/ScreenSelector.jsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        return -1;
+      } else if (b.key === defaultScreenId) {
+        return 1;
+      } else {
+        return a.key.localeCompare(b.key);
+      }
+    });
+
+    return (
+        <select
```
Comment (Bjvanminnen) [inline apps/src/applab/ScreenSelector.jsx]: extra indent

File: apps/src/applab/PlaySpaceHeaderStyles.js (hunk 1 above)
Comment (Bjvanminnen) [inline apps/src/applab/PlaySpaceHeaderStyles.js:19]: Might be worth verifying the right thing happens on different browsers here

File: apps/src/applab/PlaySpaceHeader.jsx [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+          </ToggleGroup>
+      );
+    }
+
+    if (this.state.mode === Mode.CODE && !this.props.hideViewDataButton) {
+      rightSide = <ViewDataButton onClick={this.props.onViewDataButton} />;
+    } else if (this.state.mode === Mode.DESIGN) {
+      rightSide = <ScreenSelector screenIds={this.props.screenIds}
+                                       activeScreen={this.state.activeScreen}
+                                       onChange={this.handleScreenChange} />;
```
Comment (Bjvanminnen) [inline apps/src/applab/PlaySpaceHeader.jsx]: too much indenting

File: apps/src/applab/PlaySpaceHeader.jsx [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    if (!this.props.hideToggle) {
+      leftSide = (
+          <ToggleGroup selected={this.state.mode} onChange={this.handleSetMode}>
+            <ToggleButton
+                id='codeModeButton'
+                value={Mode.CODE}>{msg.codeMode()}</ToggleButton>
+            <ToggleButton
+                id='designModeButton'
+                value={Mode.DESIGN}>{msg.designMode()}</ToggleButton>
+          </ToggleGroup>
```
Comment (Bjvanminnen) [inline apps/src/applab/PlaySpaceHeader.jsx]: too much indenting. I'm saying this a lot, which is making me wonder if perhaps we just disagree on/need to better codify when to indent 4 vs 2 spaces.

To me it just feels like a waste of horizontal space in instances like this.

Also, the ToggleButton's can be self-closing, and I think should following one of the following forms

```
// (1)
<ToggleButton
    id='codeModeButton'
    value='something/>
//(2)
<ToggleButton
    id='codeModeButton'
    value='something>
</ToggleButton>
```

i.e. if not self closing, close tag should have its own line

**Comment by Bjvanminnen** [comment]
Lots of nits, but generally LGTM. Might be worth having a discussion about indentation

File: apps/src/applab/ViewDataButton.jsx (hunk 5 above)
Comment (islemaster) [inline apps/src/applab/ViewDataButton.jsx]: Nope, this was an accident.  Will fix.

File: apps/src/applab/constants.js (hunk 3 above)
Comment (islemaster) [inline apps/src/applab/constants.js:7]: :frowning: Yes, this one is, but it's also an ID that we depend on to know that we're creating a new screen.

``` javascript
if (screenId === constants.NEW_SCREEN) {
  screenId = this.props.onScreenCreate();
}
```

We could make this check the loc'd string, but I'd rather change this to check something else.  I don't think we prevent the user from making a screen called "New screen..."...

File: apps/src/applab/ViewDataButton.jsx (hunk 4 above)
Comment (islemaster) [inline apps/src/applab/ViewDataButton.jsx]: Is an argument list a line continuation?  I assumed it was, but I'm cool if we want to treat this like a block.

File: apps/src/applab/ViewDataButton.jsx (hunk 6 above)
Comment (islemaster) [inline apps/src/applab/ViewDataButton.jsx]: In this case I'm just trying to preserve the pre-refactor behavior.  Agree that this is weird.  I'll change it up and try to verify it doesn't break anything.

File: apps/src/applab/ViewDataButton.jsx (hunk 4 above)
Comment (Bjvanminnen) [inline apps/src/applab/ViewDataButton.jsx]: Yeah, reading our style guide again, I think you probably do meet it. I think that maybe I just don't like this particular style :-P  The only time that I like a four space indent is when the line continuation precedes a block, and I want to distinguish the two different lines, for example

```
// bad
module.exports.fairlyLongFunction = function fairlyLongFunction(argumentNumber1,
  argumentNumberTwo, argumentNumber3) {
  console.log("I'm aligned with the continuation");
}

//good
module.exports.fairlyLongFunction = function fairlyLongFunction(argumentNumber1,
    argumentNumberTwo, argumentNumber3) {
  console.log("I'm not aligned with the continuation");
}
```

Anywhere else, it feels like we're just wasting horizontal space.

File: apps/src/applab/PlaySpaceHeader.jsx (hunk 12 above)
Comment (islemaster) [inline apps/src/applab/PlaySpaceHeader.jsx]: The toggle buttons aren't self-closing because we pass the button label as child content; if you strip the properties it reads like this:

``` xml
<ToggleButton>{msg.codeMode()}</ToggleButton>
```

I'd like to leave it that way for two reasons:
1. This reflects the final form the markup will take, `<button>label</button>`
2. You can pass multiple children, which would enable something like putting a font-awesome icon on the button.

Comment (Bjvanminnen) [inline apps/src/applab/PlaySpaceHeader.jsx]: Ahh yes, I misread. I would then prefer the form:

```
<ToggleButton
    id='codeModeButton'
    value='something>
  {msg.designMode()}
</ToggleButton>
```

File: apps/src/applab/ToggleGroup.jsx (hunk 7 above)
Comment (islemaster) [inline apps/src/applab/ToggleGroup.jsx]: I think so?  I wanted it to be clear that we're not rebinding `this` to the `ToggleGroup` when we call `this.props.onChange`.  Pretty sure as long as it's already bound when it's passed in as a property it should be fine (and I think it gets implicitly bound in React-land?) but I wasn't 100% sure and thought making the call in a helper made that more explicit.

File: apps/src/applab/ViewDataButton.jsx (hunk 4 above)
Comment (islemaster) [inline apps/src/applab/ViewDataButton.jsx]: Oh for [object spread properties...](https://github.com/sebmarkbage/ecmascript-rest-spread)

``` javascript
var showDataButtonStyle = {
  float: 'right',
  textAlign: 'left',
  maxWidth: '100%',
  whiteSpace: 'nowrap',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  ...styles.buttonStyle,
  ...styles.inactiveStyle
};
```

Comment (Bjvanminnen) [inline apps/src/applab/ViewDataButton.jsx]: Yep. One of my favorite ES6 features. Soon we shall have this (I hope).

Comment (Bjvanminnen) [inline apps/src/applab/ViewDataButton.jsx]: Though I guess object spread properties might be ES7

File: apps/src/applab/PlaySpaceHeader.jsx (hunk 12 above)
Comment (islemaster) [inline apps/src/applab/PlaySpaceHeader.jsx]: Does that potentially introduce whitespace inside the element, or does JSX magically clean that up for us?  I know in the past, at least writing straight-up HTML, I have to avoid newlines between an opening tag and its content if I'm being precise about the output.

I'd end up with something like this:

```
<ToggleButton
  id='codeModeButton'
  value='something'
  >{msg.designMode()}</ToggleButton>
```

Which is gross but it's to make sure I get:

```
<button>Design</button>
```

instead of

```
<button> Design </button>
```

Comment (Bjvanminnen) [inline apps/src/applab/PlaySpaceHeader.jsx]: JSX cleans up whitespace fairly aggressively. http://andrewhfarmer.com/how-whitespace-works-in-jsx/ appears to have some pretty good examples on this.

Comment (islemaster) [inline apps/src/applab/PlaySpaceHeader.jsx]: That explains a lot.  Thanks!  I'll update to your preferred format throughout.

File: apps/src/applab/designMode.js (hunk 2 above)
Comment (islemaster) [inline apps/src/applab/designMode.js:862]: Tried this out.  Our build system handles it just fine, but our linter doesn't expect to find XML elements in .js files.

Comment (Bjvanminnen) [inline apps/src/applab/designMode.js:862]: Hah. I bet we can tell our linter about JSX (though it's possible it would need an update)

File: apps/src/applab/constants.js (hunk 3 above)
Comment (islemaster) [inline apps/src/applab/constants.js:7]: FWIW, I'd like to make this change in a different PR.

**Comment by islemaster** [comment]
Interesting repeat warning while circle is running tests:

> Warning: bind(): React component methods may only be bound to the component instance. See PlaySpaceHeader

**Comment by islemaster** [comment]
@Bjvanminnen PTAL at two most recent commits.  Tests were failing because the partially-specified `ToggleButton` components passed to the `ToggleGroup` were causing props validation warnings, which (non-obviously) failed the test.  Changed so ToggleGroup accepts bare `button` children and changes them to `ToggleButton` internally, and added prop validation to make sure `ToggleGroup` only gets `button` children.

**Comment by Bjvanminnen** [comment]
lgtm
